Added a chapter about open source tools (#8)
KhronosGroup/OpenCL-Guide: chapters/os_tooling.md

# Offline Compilation of OpenCL Kernel Sources

Aside from online compilation during application execution, OpenCL kernel sources can be compiled offline into binaries that can be loaded into the drivers using special API calls (e.g. `clCreateProgramWithBinary` or `clCreateProgramWithIL`).

This section describes available open source tools for offline compilation of OpenCL kernels.

## Open Source Tools

> * [clang](https://clang.llvm.org/) is a compiler front-end for the C/C++ family of languages including OpenCL C and C++ for OpenCL that can produce executable binaries (e.g. AMDGPU), or portable binaries (e.g. SPIR). It is part of [the LLVM compiler infrastructure project](https://llvm.org/), and there is information regarding [OpenCL kernel language support and standard headers](https://clang.llvm.org/docs/UsersManual.html#opencl-features).

Reviewer: I wonder if it makes more sense to mention LLVM IR here vs. SPIR.  Note that the diagram below says "LLVM IR".

Yes, there is not much of a difference between the two in practice...

Author: Good point. Actually, I wonder if it's better to say "SPIR flavor of LLVM IR" here because technically it is no longer a conformant SPIR.

Author: FYI I have added this clarification later where individual commands are explained. Otherwise it might not be clear enough here.

* [libclc](https://libclc.llvm.org/) is a generic and portable implementation of OpenCL builtin function libraries for OpenCL 1.1 - and some functions from later versions of OpenCL can be found there too.
* [SPIRV-LLVM Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator) is a library and __llvm-spirv__ tool for translating between LLVM IR and SPIR-V. 
* [clspv](https://github.com/google/clspv) compiler and [clvk](https://github.com/kpet/clvk) runtime layer enable OpenCL applications to be executed with Vulkan drivers.
* [SPIR-V Tools](https://github.com/KhronosGroup/SPIRV-Tools) provide a set of utilities to process SPIR-V binaries including __spirv-opt__ optimizer, __spirv-link__ linker, __spirv-dis__/__spirv-as__ (dis-)assembler, and __spirv-val__ validator.

### Compiling Kernels to SPIR-V

> Available open source tools can be used to perform full compilation from OpenCL kernel sources into SPIR-V.

Reviewer 2: ```suggestion
The open source tools can be used to perform full compilation from OpenCL kernel sources into SPIR-V.
```

<p align="center">
<br>
<img src="../images/opencl_to_spirv_tooling.jpg">
<br> <br>
  <b>Offline compilation flow for OpenCL kernels into SPIR-V</b>
<br> <br>
</p>

#### Examples

If you want to try the above compilation flow for yourself, after installing the tools, you can use the following commands.

##### Compile for OpenCL runtime

__(i)__ Compiling OpenCL C/C++ for OpenCL file into intermediate LLVM IR (for 32 bit targets) formats:

```

> clang -c -target spir -O0 -emit-llvm -o test.bc test.cl

Reviewer: Does it make sense to have some Compiler Explorer links to demonstrate these command lines?

Author: Yes this is a good idea. The only issue is that the exact command-lines won't be the same since Compiler Explorer doesn't support OpenCL natively. :( 

So we could need to add -x flag everywhere. Do you think it is ok?

Author: I will resolve this for now. We can always add those in a separate PR. :)

```
Using `-cl-std` changes the language version compiled for.

> To compile C++ for OpenCL source pass `-cl-std=CLC++` or use the following file extension `.clcpp`.

Reviewer 2: ```suggestion
To compile C++ for OpenCL source pass `-cl-std=CLC++` or use the file extension `.clcpp`.
```

```
clang -cl-std=CLC++ -c -target spir -O0 -emit-llvm -o test.bc test.cl
clang -c -target spir -O0 -emit-llvm -o test.bc test.clcpp
```
If debugging support is needed then the `-g` flag can be used in the clang invocation:

```
clang -c test.cl -target spir -o test.bc -g
```

> __Note:__ In clang releases up to 12.0, calling most builtin functions requires an extra flag to be passed to clang `-Xclang -finclude-default-header`. Refer to the release documentation for more details.

Reviewer 2: ```suggestion
__Note:__ In clang releases up to 12.0, calling most builtin functions requires the extra flags `-Xclang -finclude-default-header` to be passed to clang. Refer to the release documentation for more details.
```

> Refer to the release documentation for more details.

Maybe provide a link?

__(ii)__ Converting LLVM IR into SPIR-V. 

```
llvm-spirv test.bc -o test.spv
```

> __Note:__ Converting optimized IR is currently unsupported. However, the stand-alone __spirv-opt__ tool can be used to optimize SPIR-V modules once they have been created.

Reviewer: I think this is a little too strong.  I frequently convert optimized IR, and there are nice advantages in doing so (when it works).

Author: Yes, we do support many features but then some we don't. My worry about advertising it officially is that we still haven't agreed within the dev community that we support this functionality (https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/203). Hence if developers do create bugs related to this we don't guarantee to fix those or we don't guarantee how quickly they will be fixed. So I just want to avoid disappointments. But perhaps we can change the wording to say that this is an **experimental** functionality i.e. expected to work and if not bugs might be fixed but not guaranteed to be fixed.

Author: We have discussed it on the Tooling TSG call on Apr 1 and I have captured our decision in the last update.

I would like @svenvh and @AlexeySachkov to proofread this before finalizing the PR.

Reviewer 3: The new wording of this paragraph looks good to me

> # Offline Compilation of OpenCL Kernel Sources

Reviewer 4: Would it be simpler to call this page `Offline Compilation of OpenCL Kernel Sources to SPIR-V`.

Compiling to binaries only gets 2-3 sentences, the rest is about SPIR-V.   I'm not sure how to write more about compiling to binaries as it's all vendor specific tools.

This is just an option to consider.

Author: I think I would prefer to leave it more general. For example, clang can compile to some binaries i.e. AMDGPU and I am linking in a page where it is described although I appreciate it is pretty trivial.

Reviewer 4: Okay.

> Aside from online compilation during the application execution, OpenCL kernel sources can be compiled offline into binaries that can be loaded into the drivers using special API (e.g. `clCreateProgramWithBinary` or `clCreateProgramWithIL`).

Reviewer 4: ```suggestion
Aside from online compilation during application execution, OpenCL kernel sources can be compiled offline into binaries that can be loaded into the drivers using special API calls (e.g. `clCreateProgramWithBinary` or `clCreateProgramWithIL`).
```

This section described available open source tools for offline compilation of OpenCL kernels.

## Open Source Tools

> * [clang](https://clang.llvm.org/) is a compiler front-end for the C/C++ family of languages including OpenCL C and C++ for OpenCL that can compile into an executable binary (e.g. AMDGPU), or a portable binary (e.g. SPIR). It is part of [the LLVM compiler infrastructure project](https://llvm.org/), and there is information regarding [OpenCL kernel language support and standard headers](https://clang.llvm.org/docs/UsersManual.html#opencl-features).

Reviewer 4: ```suggestion
* [clang](https://clang.llvm.org/) is a compiler front-end for the C/C++ family of languages, including OpenCL C and C++ for OpenCL. It can compile into an executable binary (e.g. AMDGPU), or a portable binary (e.g. SPIR). It is part of [the LLVM compiler infrastructure project](https://llvm.org/), and there is information regarding [OpenCL kernel language support and standard headers](https://clang.llvm.org/docs/UsersManual.html#opencl-features).
```

Author: Added manually due to rebasing issue.

> * [libclc](https://libclc.llvm.org/) is a generic and portable implementation of OpenCL builtin function libraries for OpenCL 1.1 - and some functions from later versions of OpenCL can be found there too.

Reviewer 4: libclc feels like an odd one out in the list.  All the others are something that is useful to an end-user programmer, libclc is useful for tool implementers.   Maybe just drop it from this list?

Author: The only relevant use case I could think of is clspv.

But perhaps you are right it would be most likely hidden behind the compiler driver setup.

I am looping in @alan-baker in case he has some more input.

Otherwise, we can remove it indeed.

Reviewer 5: My thought was: where is libclc in the image? I would either add it as a box and hook it up to the linking part of both sides (to communicate that implementations of built-in have to be provided by some entity and don't materialize from thin air), or either drop it from the text.

It's either important enough for people to get the "whole picture" or not, but not both.

Reviewer: I think we should try to figure out how to make libclc more useful and accessible, but I agree that it doesn't add much to this list right now.

Author: I have removed it for now. This can be added any time we find a good use case for the end-user.

> * [SPIRV-LLVM Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator) is a library and __llvm-spirv__ tool for translating between LLVM IR and SPIR-V. 

Author: I would like to mention llvm-spirv since it appears in the diagram. How about this:

```suggestion
* [SPIRV-LLVM Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator) provides a library and __llvm-spirv__ tool for bidirectional translation between LLVM IR and SPIR-V. 
```

> * [SPIRV-LLVM Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator) is a library and __llvm-spirv__ tool for translating between LLVM IR and SPIR-V. 

Reviewer 2: Add missing "the":

```suggestion
* [SPIRV-LLVM Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator) provides a library and the __llvm-spirv__ tool for bidirectional translation between LLVM IR and SPIR-V. 
```

* [clspv](https://github.com/google/clspv) compiler and [clvk](https://github.com/kpet/clvk) runtime layer enable OpenCL applications to be executed with Vulkan drivers.
* [SPIR-V Tools](https://github.com/KhronosGroup/SPIRV-Tools) provide a set of utilities to process SPIR-V binaries including __spirv-opt__ optimizer, __spirv-link__ linker, __spirv-dis__/__spirv-as__ (dis-)assembler, and __spirv-val__ validator.

### Compiling Kernels to SPIR-V

Available open source tools can be used to perform full compilation from OpenCL kernel sources into SPIR-V.

<p align="center">
<br>

> <img src="../images/opencl_to_spirv_tooling.jpg">

Author: > A nit about the image: I would only mention .clcpp, not .cl/.clcpp, for C++ for OpenCL, to encourage its adoption. This is of course conditional on the corresponding PR in LLVM. Unless we consider this might be too new to be available?

I think you make a good point and the change was already merged. In the text I do have another option using _.cl_ extension so I think it is fine to keep _.clcpp_ only in the diagram. This should be the flow we advertise anyway.

> A possible point of confusion: the two flavors of SPIR-V are not mentioned in this document, nor in the figure. It might be misleading in that someone could think a Vulkan SPIR-V module can be linked with an OpenCL SPIR-V module. I'm not sure how important that is, in the context of this document.

Yeah, I agree we should avoid such confusion. So there are two changes I could think of:
1. We can add a footnote in the diagram for `spirv-link` box explaining that it is not possible to link modules from two dialects
2. We can just add a note in the text of `spirv-link` example.

Any preferences or other suggestions?

> <img src="../images/opencl_to_spirv_tooling.jpg">

Reviewer 5: > A possible point of confusion: the two flavors of SPIR-V are not mentioned in this document...

Having read the markdown version before discussions, I would like to second that. Not just the linking part, but the image may be interpreted that Vulkan compute shaders could be authored using C++ for OpenCL and use the OpenCL flavor of the toolchain to compile them to "just" SPIR-V (as opposed to compute SPIR-V).

> <img src="../images/opencl_to_spirv_tooling.jpg">

Author: > Vulkan compute shaders could be authored using C++ for OpenCL

Just to make sure I understood you. Do you mean that the figure might be misleading someone to think that Vulkan shaders can be written in C++ for OpenCL?

>  and use the OpenCL flavor of the toolchain to compile them to "just" SPIR-V (as opposed to compute SPIR-V).

Do you suggest that we should indicate somehow on the diagram that clspv path uses Vulkan flavour SPIR-V and the other path uses OpenCL flavour SPIR-V? Or do you think it is sufficient to explain it in the text?

> <img src="../images/opencl_to_spirv_tooling.jpg">

Reviewer 6: > Any preferences or other suggestions?

To be honest, I'm not sure what the right solution is -- and I'm personally happy if this is dealt with at a later time. It might be better to refer explicitly to each SPIR-V flavor in the figure to avoid any confusion. But at the same time I was surprised that the documentation of the SPIR-V Tools doesn't mention the incompatibility of the two flavors so maybe a more global clarification would be useful.

> <img src="../images/opencl_to_spirv_tooling.jpg">

Author: > It might be better to refer explicitly to each SPIR-V flavor in the figure to avoid any confusion.

This can be done in a top part but then there is a common part and I don't have a good way to achieve this apart from just drawing the teo flows completely separtely. Does it make sense?

Otherwise, I will explain it in the text for now. Actually, I think it should be explained in the text anyway. Then we can see if the diagram needs to be updated or we can postpone it and solve in follow up PRs.

> <img src="../images/opencl_to_spirv_tooling.jpg">

Author: I have added a paragraph below the diagram. Do you think it is good enough for now?

> <img src="../images/opencl_to_spirv_tooling.jpg">

Reviewer 6: > I have added a paragraph below the diagram. Do you think it is good enough for now?

Looks good to me.

<br> <br>
  <b>Offline compilation flow for OpenCL kernels into SPIR-V</b>
<br> <br>
</p>

#### Examples

If you want to try the above compilation flow for yourself, after installing the tools, you can use the following commands.

##### Compile for OpenCL runtime

__(i)__ Compiling OpenCL C/C++ for OpenCL file into intermediate LLVM IR (for 32 bit targets) formats:

```
clang -c -target spir -O0 -emit-llvm -o test.bc test.cl
```
Using `-cl-std` changes the language version compiled for.

To compile C++ for OpenCL source pass `-cl-std=CLC++` or use the following file extension `.clcpp`.

```
clang -cl-std=CLC++ -c -target spir -O0 -emit-llvm -o test.bc test.cl
clang -c -target spir -O0 -emit-llvm -o test.bc test.clcpp
```

> If debugging support is needed  can be obtained `-g` flag can be passed in clang invocation:

Reviewer 4: ```suggestion
If debugging support is needed then the `-g` flag can be used in the clang invocation:
```

```
clang -c test.cl -target spir -o test.bc -g
```

__Note:__ In clang releases up to 12.0, calling most builtin functions requires an extra flag to be passed to clang `-Xclang -finclude-default-header`. Refer to the release documentation for more details.

__(ii)__ Converting LLVM IR into SPIR-V. 

```
llvm-spirv test.bc -o test.spv
```

> __Note:__ converting optimized IR is currently unsupported. Therefore, a stand-alone __spirv-opt__ tool can be used to optimize SPIR-V modules.

Reviewer 4: ```suggestion
__Note:__ Converting optimized IR is currently unsupported. However, the stand-alone __spirv-opt__ tool can be used to optimize SPIR-V modules once they have been created.
```

> __Note:__ Converting IR produced with optimization levels other than `-O0` is only available as an experimental feature and it is not guaranteed to work. The majority of common functionality is expected to convert even when optimizations are enabled. Developers are encouraged to file a bug report when issues are encountered. When translating modules obtained with optimizations can not be not supported, the stand-alone __spirv-opt__ tool can be used to optimize in SPIR-V format once the conversion is done from LLVM IR produced with optimizations disabled.

Reviewer 2: Looks good, just rewording a little to make it clearer for the reader:

```suggestion
__Note:__ Converting IR produced with optimization levels other than `-O0` is only available as an experimental feature and is not guaranteed to work. The conversion should work for the majority of IR constructs even when optimizations are enabled, but certain IR constructs may cause issues. Developers are encouraged to file a bug report when issues are encountered. As a workaround when encountering an issue in translating modules obtained with optimizations, generate LLVM IR with optimizations disabled, and then use the stand-alone __spirv-opt__ tool to optimize at the SPIR-V level.
```

Author: For the developers I prefer not to tight the description to the IR constructs since they are not supposed to know details about it.

Author: ```suggestion
__Note:__ Converting IR produced with optimization levels other than `-O0` is only available as an experimental feature and it is not guaranteed to work. The majority of common functionality is expected to convert even when optimizations are enabled. Developers are encouraged to file a bug report when issues are encountered. As a workaround when encountering an issue in translating modules obtained with optimizations, generate LLVM IR with optimizations disabled, and then use the stand-alone __spirv-opt__ tool to optimize at the SPIR-V level.
```

Author: How about we keep the first part as is?

Reviewer 2: The words "common functionality" in combination with "is expected to convert" is unclear to me.  Maybe I misunderstood what you are trying to say with "common functionality" in this context?

Regardless, I think you should still rephrase "expected to convert" into "conversion [...] is expected to succeed".

Author: Ok, I have changed to

` In the majority of cases, the conversion is expected to succeed when optimizations are enabled.`

Does it make sense?

Reviewer 2: Yes, that's better, thanks!

__(iii)__ Linking multiple modules can be done on LLVM IR level by passing multiple `.bc` files to __llvm-link__.

```
llvm-link test1.bc test2.bc -o app.bc
llvm-spirv app.bc -o app.spv
```
Alternatively, __spirv-link__ can be used:

```
spirv-link -o app.spv test1.spv test2.spv
```

__(iv)__ Once the SPIR-V binary is produced, it can be loaded in OpenCL applications using the `clCreateProgramWithIL` API call.

##### Compile for Vulkan runtime

Compiling OpenCL sources directly to SPIR-V binary format:

```
clspv test.cl -o test.spv
```

> The SPIR-V binary files can be further optimized or linked using __spirv-opt__ and __spirv-link__ respectively before loading into Vulkan runtime (TODO: API call).

Reviewer 4: ```suggestion
The SPIR-V binary files can be further optimized or linked using __spirv-opt__ and __spirv-link__ respectively before loading into the Vulkan runtime using the `vkCreateShaderModule` API call.
```
